File: LeetCode/Medium/0006_Zigzag_Conversion/Solution.cpp
// 2023-02-02
class Solution {
public:
    string convert(string s, int numRows) {
        /*
            input: a string, and an int number of rows
            output: the resulting zigzag pattern read line by line

            notes:
            - looks pretty simple; initialize vector<vector<char>> of size numRows
            - add character by character to vector 0, 1, 2, ...numRows-1, numRows-2, ... 1, 0
              and repeat until we run out of characters
            - append characters from each row togetehr into strings, then append strings together

            analysis:
            - time = O(n) where n=length of s
            - space = O(n) where n=length of s
        */
        if (numRows == 1)
            return s;

        int n = s.length();
        vector<string> rows(numRows);
        int row = 0;
        for (int i=0; i<n; i++) {
            rows[row] += s[i];
            if ((i/(numRows-1)) & 1 == 1) {
                row--;
            } else {
                row++;
            }
        }

        string output = "";
        for (string& row : rows) {
            output += row;
        }
        return output;
    }
};

File: LeetCode/Medium/0006_Zigzag_Conversion/Solution.java
// 2021-06-08
class Solution {
    public String convert(String s, int numRows) {
        
        // Base case of single row
        if (numRows == 1) return s;
        
        // Create array that holds StringBuilder for each row
        StringBuilder[] rows = new StringBuilder[numRows];

        // Initialize array of StringBuilder
        for (int i = 0; i < rows.length; i++){
            rows[i] = new StringBuilder();
        }
        
        // Character position, whether we are zigging or zagging, and the row number
        int count = 0;
        int change = 1;
        int rowNo = 0;
        
        while (count < s.length()) {
            
            // Add character into designated StringBuilder position
            rows[rowNo].append(s.charAt(count));
            
            // Change zigzag direction
            if (rowNo + change == numRows || rowNo + change < 0) {
                change = -change;
            }
            count++;
            rowNo += change;
        }
        
        // Append all together
        for (int i = 1; i < rows.length; i++) {
            rows[0].append(rows[i]);
        }
        
        return rows[0].toString();
    }
}